File: flutter/lib/domain/space/coord_2d.dart
import 'package:ship_conquest/domain/space/position.dart';

/// Represents a 2D coordinate with integer values.
class Coord2D {
  final int x;
  final int y;

  Coord2D({required this.x, required this.y});

  @override
  int get hashCode => Object.hash(x, y);

  //functions to operate with Coord2D instances
  Coord2D operator -(Coord2D other) => Coord2D(x: x - other.x, y: y - other.y);
  Coord2D operator +(Coord2D other) => Coord2D(x: x + other.x, y: y + other.y);
  Position operator *(double other) => Position(x: x * other, y: y * other);

  @override
  bool operator == (Object other)  =>
      other is Coord2D && other.x == x && other.y == y;

  @override
  String toString() => 'Coord(x = $x, y = $y)\n';
  Position toPosition() => Position(x: x.toDouble(), y: y.toDouble());
}

File: jvm/ship-conquest/src/main/kotlin/com/example/shipconquest/domain/Coord2D.kt
package com.example.shipconquest.domain

data class Coord2D(val x: Int, val y: Int) {
    operator fun plus(position: Coord2D) =
        Coord2D(x = this.x + position.x, y = this.y + position.y)

    operator fun minus(position: Coord2D) =
        Coord2D(x = this.x - position.x, y = this.y - position.y)
}

fun Coord2D.toPosition(): Position = Position(x = this.x.toDouble(), y = this.y.toDouble())